### PR TITLE
[kernel-spark] Enable a few E2E dsv2 streaming tests and fix offset management edge cases 

### PR DESCRIPTION
**Pull request by zikangh**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta/pull/5652/files) to review incremental changes.
- [**stack/integration2**](https://github.com/delta-io/delta/pull/5652) [[Files changed](https://github.com/delta-io/delta/pull/5652/files)]
  - [stack/reader](https://github.com/delta-io/delta/pull/5638) [[Files changed](https://github.com/delta-io/delta/pull/5638/files/52a3c8381bce050a392701b2a85b6a149b7ca9d3..5c9af512429c6f83c55538cff64c4c75502e1c12)]
    - [stack/lazy](https://github.com/delta-io/delta/pull/5650) [[Files changed](https://github.com/delta-io/delta/pull/5650/files/5c9af512429c6f83c55538cff64c4c75502e1c12..cce1148828608ac023bb94d5a5b0f5c43fe4cc25)]
      - [stack/lazy2](https://github.com/delta-io/delta/pull/5686) [[Files changed](https://github.com/delta-io/delta/pull/5686/files/cce1148828608ac023bb94d5a5b0f5c43fe4cc25..c4059b4696f87ed38fcfc43ef19ba65ae090f517)]
        - [stack/snapshot](https://github.com/delta-io/delta/pull/5651) [[Files changed](https://github.com/delta-io/delta/pull/5651/files/c4059b4696f87ed38fcfc43ef19ba65ae090f517..4d3946935a87170d58a2b219431ed536c979cce5)]

---------

#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description

  Add integration tests for kernel-spark DSv2 streaming and fix edge cases in offset handling.

  **Changes:**

  1. **Fixed offset capping in `getBatch()`**: Cap `endVersion` to the latest available snapshot version to handle offsets pointing to non-existent future versions (which can occur when
  `buildOffsetFromIndexedFile` advances to the next version).

  2. **Fixed starting version caching**: Made `getStartingVersion()` synchronized and added caching to ensure idempotent behavior - `startingVersion: "latest"` must resolve to the same version across
  multiple calls.

  4. **Created `DeltaSourceDSv2Suite`**: Runs a subset of existing DSv1 streaming tests using DSv2 to verify API compatibility for `startingVersion` options.

## How was this patch tested?

End-to-end streaming tests. 

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java [hunk 1]
```diff
@@ -333,15 +337,20 @@ public void stop() {
    *
    * <p>This is the DSv2 Kernel-based implementation of DeltaSource.getStartingVersion.
    */
-  Optional<Long> getStartingVersion() {
+  synchronized Optional<Long> getStartingVersion() {
```
Comment (gengliangwang) [inline kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:388]: We can make it more efficient
```
private volatile Optional<Long> cachedStartingVersion = null;

Optional<Long> getStartingVersion() {
    Optional<Long> cached = cachedStartingVersion;  // Single volatile read
    if (cached != null) {
        return cached;
    }
    synchronized (this) {
        // Double-check after acquiring lock
        if (cachedStartingVersion != null) {
            return cachedStartingVersion;
        }
        // ... compute result ...
        cachedStartingVersion = result;
        return result;
    }
}
```

Comment (zikangh) [reply kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:388]: Thank you. Perhaps this is overkill? We only call getStartingVersion() once per stream.

Comment (gengliangwang) [reply kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:388]: yeah this is a minor comment. I am fine with the current implementation

**Review by gengliangwang — APPROVED** [review]

**Review by huan233usc — APPROVED** [review]

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala [hunk 2]
```diff
@@ -1315,6 +1315,89 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase
     }
   }
 
+  test("new commits arrive after stream initialization") {
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: why do you need a new test? did the existing tests that have been around for years not cover this particular case?

Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: whats the difference between this test and next one? initialization with latest version vs startingVersion? if so, please make it clear in the name.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.delta.test.V2ForceTest
+
+/**
+ * Test suite that runs DeltaSourceSuite using the V2 connector (V2_ENABLE_MODE=STRICT).
+ */
+class DeltaSourceDSv2Suite extends DeltaSourceSuite with V2ForceTest {
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: @huan233usc how long are we going to continue this pattern of overriding individual tests liek this?

if we want to add such tests in the short term, what is the long term plan of these v2 only tests when we will be running everything in v2 mode?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: can we put all teh dsv2 specific tests in a separate directory so that we know which ones are dsv2 only tests?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ */
+class DeltaSourceDSv2Suite extends DeltaSourceSuite with V2ForceTest {
+
+  override protected def useDsv2: Boolean = true
+
+  /**
+   * Override disableLogCleanup to use DeltaLog API instead of SQL ALTER TABLE.
+   * Path-based ALTER TABLE doesn't work properly with V2_ENABLE_MODE=STRICT.
+   */
+  override protected def disableLogCleanup(tablePath: String): Unit = {
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: @huan233usc how are we tracking these gaps ?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  private lazy val testsToRun = Set(
+    "startingVersion",
+    "startingVersion latest",
+    "startingVersion latest defined before started",
+    "startingVersion latest works on defined but empty table",
+    "new commits arrive after stream initialization with startingVersion"
+  )
+
+  override protected def shouldSkipTest(testName: String): Boolean = {
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: @huan233usc can we change this from "should skip" to "expected to fail"? 
basically, if the status of some tests changes for some reason from fail --> pass, but its fully skipped, how will we catch those changes?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    )
+  }
+
+  private lazy val testsToRun = Set(
+    "startingVersion",
+    "startingVersion latest",
+    "startingVersion latest defined before started",
+    "startingVersion latest works on defined but empty table",
+    "new commits arrive after stream initialization with startingVersion"
+  )
```
Comment (tdas) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: i think we should start with making a list of tests that should skipped so that we know exactly what is broken. its hard to understand otherwise how much is broken and needs more work.

File: kernel-spark/src/test/java/io/delta/kernel/spark/Dsv2BasicTest.java [hunk 8]
```diff
@@ -164,46 +171,148 @@ public void testColumnMappingRead(@TempDir File deltaTablePath) {
   }
 
   @Test
-  public void testStreamingRead(@TempDir File deltaTablePath) {
+  public void testStreamingReadMultipleVersions(@TempDir File deltaTablePath) throws Exception {
```
Comment (tdas) [inline kernel-spark/src/test/java/io/delta/kernel/spark/Dsv2BasicTest.java:174]: @huan233usc what is the scope of the Dsv2BasicTest class? are all operations supposed to be added there? i want some structure and scope to each test suite, so that we dont end up with randomization in what test is where. 

why not make a separate test suite for streaming?

Comment (huan233usc) [reply kernel-spark/src/test/java/io/delta/kernel/spark/Dsv2BasicTest.java:174]: Yes, I think we should separate the test suites for streaming, and probably write and DML in the future. 

However, for those tests kernel-spark/src/test/java/io/delta/kernel/spark/* I was thinking about consolidate it with existing v1 tests(which now could run with both v1 and v2 connector now) in the end

Dsv2BasicTest was added before we made the module change to enable running v1 and v2 connector with DeltaCatalog. It has a setup that we could write using v1 connector but read with v2 connector(with a special test catalog impl io.delta.kernel.spark.catalog.TestCatalog), and so we could cover some simple e2e scenario in the very beginning.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 4 above)
Comment (huan233usc) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: +1 to `spark-unified/tests/v2 folder` for example

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 6 above)
Comment (huan233usc) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Good point, "expected to fail" is a better pattern than "should skip".

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 5 above)
Comment (huan233usc) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Is it the general missing of `alter table` ddl support. We could track it via a github issue with a tag describing it is a v2 connector FR etc.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 3 above)
Comment (huan233usc) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: For streaming development, I think we will have this pattern for DeltaSource*Suite

Meanwhile I was thinking about replacing this allowlisting model with denylisting model (setting a github workflow with "auto" v2 selection) The resource might be a concern.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 4 above)
Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Moved it to `spark-unified/src/test/` so it sits next to `V2ForceTest.scala`. Not sure if I should create the `v2/` subfolder yet. perhaps @huan233usc can decide once tests are consolidated.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala (hunk 2 above)
Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: You are right! This is mostly similar to "basic", I removed it. The other test `new commits arrive after stream initialization - with explicit startingVersion` is not redundant and provides coverage for: 
1. the startingVersion is a numeric value
2. multi-batch
3. commits come in after stream initialization

Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala]: Yes, default behavior (reads an initial snapshot) vs an explicit startingVersion. I renamed one of the test cases to disambiguate.

File: kernel-spark/src/test/java/io/delta/kernel/spark/Dsv2BasicTest.java (hunk 8 above)
Comment (zikangh) [reply kernel-spark/src/test/java/io/delta/kernel/spark/Dsv2BasicTest.java:174]: +1 we should consolidate eventually; currently DSv2BasicTest uses a custom TestCatalog, independent from DeltaCatalog's V2_ENABLE_MODE routing. imo it does provide value right now as we can test dsv2 without having to worry about integration/routing bugs.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 7 above)
Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Done. also made a list of shouldPass tests & added validation (a test case should either in shouldPass or shouldFail).

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 6 above)
Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Done.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala (hunk 5 above)
Comment (zikangh) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaSourceDSv2Suite.scala]: Created https://github.com/delta-io/delta/issues/5731 and added a TODO.
@huan233usc -- could you help add details?

File: kernel-spark/src/test/scala/io/delta/kernel/spark/snapshot/unitycatalog/UCUtilsSuite.scala [hunk 9]
```diff
@@ -164,13 +164,7 @@ class UCUtilsSuite extends SparkFunSuite with SharedSparkSession {
         info.getTablePath == TABLE_PATH_ALPHA,
         s"Table path mismatch: got ${info.getTablePath}")
       assert(info.getUcUri == UC_URI_ALPHA, s"UC URI mismatch: got ${info.getUcUri}")
-      val configMap = info.getAuthConfig
```
Comment (huan233usc) [inline kernel-spark/src/test/scala/io/delta/kernel/spark/snapshot/unitycatalog/UCUtilsSuite.scala:167]: Why those changes are needed?

File: spark-unified/src/test/scala/org/apache/spark/sql/delta/test/V2ForceTest.scala [hunk 10]
```diff
@@ -44,15 +44,15 @@ trait V2ForceTest extends DeltaSQLCommandTest {
...
-   * Override `test` to apply the `shouldSkipTest` logic.
-   * Tests that should be skipped are converted to ignored tests.
+   * Override `test` to apply the `shouldFail` logic.
+   * Tests that are expected to fail are converted to ignored tests.
    */
   abstract override protected def test(
       testName: String,
       testTags: Tag*)(testFun: => Any)(implicit pos: Position): Unit = {
-    if (shouldSkipTest(testName)) {
+    if (shouldFail(testName)) {
```
Comment (huan233usc) [inline spark-unified/src/test/scala/org/apache/spark/sql/delta/test/V2ForceTest.scala:53]: Can you file an issue and assign it to me to assert on test failure than just ignore the test?

File: kernel-spark/src/test/scala/io/delta/kernel/spark/snapshot/unitycatalog/UCUtilsSuite.scala (hunk 9 above)
Comment (zikangh) [reply kernel-spark/src/test/scala/io/delta/kernel/spark/snapshot/unitycatalog/UCUtilsSuite.scala:167]: Reverted.

File: spark-unified/src/test/scala/org/apache/spark/sql/delta/test/V2ForceTest.scala (hunk 10 above)
Comment (zikangh) [reply spark-unified/src/test/scala/org/apache/spark/sql/delta/test/V2ForceTest.scala:53]: Done.

**Review by tdas — APPROVED** [review]
LGTM. but some of the test set up will have to change as already discussed with @huan233usc but I want to merge this to unblock the stack of PRs after this.

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java [hunk 11]
```diff
@@ -381,15 +385,20 @@ public void stop() {
...
```
Comment (jerrypeng) [inline kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:388]: A couple of questions:
1. Why does this need to be synchronized?  What threads do we expect this method to be called from?
2. Why is the caching logic needed?  There is no such logic in the DSv1 implementation.

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java [hunk 12]
```diff
@@ -544,6 +555,20 @@ private CloseableIterator<IndexedFile> filterDeltaLogs(
     Optional<Long> endVersionOpt =
         endOffset.isPresent() ? Optional.of(endOffset.get().reservoirVersion()) : Optional.empty();
 
+    // Cap endVersion to the latest available version. The Kernel's getTableChanges requires
```
Comment (jerrypeng) [inline kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:558]: Is this because for "latest" we do

```case StartingVersionLatest =>
          deltaLog.update(catalogTableOpt = catalogTableOpt).version + 1
```
?

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java (hunk 11 above)
Comment (zikangh) [reply kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:388]: DSv1 uses lazy val: 
```
protected lazy val getStartingVersion: Option[Long] = {
    if (options.startingVersion.isDefined) {
      val v = options.startingVersion.get match {
...
```




  - Automatically provides thread-safe lazy initialization
  - Computes the value once and caches it automatically

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java (hunk 12 above)
Comment (zikangh) [reply kernel-spark/src/main/java/io/delta/kernel/spark/read/SparkMicroBatchStream.java:558]: Not quite. please see https://github.com/delta-io/delta/pull/5652/files#diff-aaaa05f9a58f7e60d07b3f257a2dabcbde77996f91bf182c0584b35b7d0930d9R563-R568
